snuba/datasets/processors/replays_processor.py

```diff
@@ -188,25 +189,78 @@ def process_message(
-            # # At time of writing, there are no reads of the values in the `processed`
-            # # dictionary to inform values in other functions.
-            # # Ideally we keep continue that rule
-            self._process_base_replay_event_values(processed, replay_event)
-            self._process_tags(processed, replay_event)
-            self._process_sdk(processed, replay_event)
-            self._process_kafka_metadata(metadata, processed)
-
-            # # the following operation modifies the event_dict and is therefore *not* order-independent
-            self._process_user(processed, replay_event)
-            self._process_event_hash(processed, replay_event)
-            self._process_contexts(processed, replay_event)
-            return InsertBatch([processed], None)
+            if replay_event["type"] == "replay_actions":
+                actions = process_replay_actions(replay_event, processed, metadata)
+                return InsertBatch(actions, None)
+            else:
+                # The following helper functions should be able to be applied in any order.
+                # At time of writing, there are no reads of the values in the `processed`
+                # dictionary to inform values in other functions.
+                # Ideally we keep continue that rule
+                self._process_base_replay_event_values(processed, replay_event)
+                self._process_tags(processed, replay_event)
+                self._process_sdk(processed, replay_event)
+                self._process_kafka_metadata(metadata, processed)
+
+                # the following operation modifies the event_dict and is therefore *not*
+                # order-independent
+                self._process_user(processed, replay_event)
+                self._process_event_hash(processed, replay_event)
+                self._process_contexts(processed, replay_event)
+                return InsertBatch([processed], None)
         except Exception:
             metrics.increment("consumer_error")
             raise
 
 
+def process_replay_actions(
+    payload: Mapping[Any, Any],
+    processed: Mapping[Any, Any],
+    metadata: KafkaMessageMetadata,
+) -> list[dict[str, Any]]:
+    """Process replay_actions message type."""
+    return [
+        {
+            # Primary-key.
+            "project_id": processed["project_id"],
+            "timestamp": default(
+                lambda: datetime.now(timezone.utc),
+                maybe(to_datetime, click["timestamp"]),
+            ),
+            "replay_id": to_uuid(payload["replay_id"]),
+            "segment_id": None,
+            "event_hash": click["event_hash"],
+            # Default values for non-nullable columns.
+            "trace_ids": [],
+            "error_ids": [],
+            "urls": [],
+            "title": None,
+            "platform": "javascript",
+            "user": None,
```

Comment:
just thinking ahead if we end up trying to go the subquery route it would be less headache if we stored all of the metadata on these rows as well, but can also be solved with a second query and joining so i think its fine as is, e.g. for the case of

"show me replays where it was user X and they clicked Y". i'm guessing we did this to save data and it makes sense.

Reply:
Unfortunately we don't have that information.  We would have to read on ingest which has two problems:

1. It doesn't scale.
2. The row is not guaranteed to exist in Clickhouse.

Merging the two events into one in Relay and/or the SDK will solve this problem as we would have both the event and the segment-data.